R/generate_projections.R

```diff
@@ -0,0 +1,207 @@
+#
+#                       IS IN
+#
+# plot_real_time_projection()
+
+# real_time_projection() -> get_mspe()
+#
+#                    IS IN
+#
+# evaluate_real_time_projection()
+
+
+#### inverse Z score #####
+#' @description Inversely transform z-score standardized data to raw data
+#'              using: raw = z*sd + mean
+#'
+#' @param old_data: the previous data that is used for z-score standardization
+#' @param scaled_data: current scaled data
+#' @return inversely z-score transformed data
+inv_z_scale <- function(old_data, scaled_data) {
+    sd <- sd(old_data)
+    m <- mean(old_data)
+
+    scaled_data * sd + m
+}
+
+
+#' @description Data format preparation for model building.
+#' It includes:
+#' 1. Add `previous` predictor.
+#' 2. Z-score standardization of `previous`,`elapsed`, and `daily`
+#' @param data: the data frame need to be prepared.
+#' Must include columns:`previous`,`elapsed`, and `daily`
+#' @return Data frame with scaled `previous`,`elapsed`, and `daily` and ready to build model
+data_prepare <- function(data) {
+    data$previous <- c(NA, data$daily[-nrow(data)]) # nolint
+    data <- data[-1, ]
+    data$previous_std <- scale(data$previous, center = T, scale = T)
+    data$elapsed_std <- scale(data$elapsed, center = T, scale = T)
+    data$daily_std <- scale(data$daily, center = T, scale = T)
+
+    return(data)
+}
+
+#' @description This function uses `train_data` to train the linear regression
+#' and project to the last date in the `test_data`.
+#' It is embedded in `plot_real_time_projection` and 'evaluate_real_time_projection'
+#' @param train_data: data frame used to train linear regression
+#' @param test_data: data frame used to compare with the prediction generated by linear regression
+#' @return a list containing the predicted uptake and the corresponding
+#' observed rate under the `test_data` time frame,
+#' in both uptake rate `rate` and cumulative uptake `cumulative`
+
+real_time_projection <- function(
+    train_data = nis_usa_2022,
+    test_data = nis_usa_2023) {
+    # transform the train data and test data to the format of model building and model prediction #
+    scaled_train <- data_prepare(train_data)
+    scaled_test <- data_prepare(test_data)
+
+    # The first row is also removed for raw test data, to be comparable with the predicted values #
+    test_data <- test_data[-1, ]
+
+    # model building #
+    model <- build_projection_model2(data = scaled_train)
+
+    ## get out-of-sample prediction ##
+    scaled_pred <- brms::posterior_predict(model, newdata = scaled_test)
+
+    ## recover the prediction to the original format for comparision ##
+    pred <- inv_z_scale(test_data$daily, scaled_pred)
+
+    # predicted daily vaccine uptake #
+    pred_summary <- data.frame(
+        date = test_data$date,
+        obs = test_data$daily,
+        lower = apply(pred, 2, quantile, 0.025),
+        upper = apply(pred, 2, quantile, 0.975),
+        mean = colMeans(pred)
```

Comment:
This is the per-day average incident uptake over the time period between data points (typically 7 days). It must be multiplied by the time interval since the previous data point (in days) to be the actual incident uptake.

Reply:
Daily prediction is used in the python version.

Reply:
But why *don't* we do weekly? We'll never have daily data, so why are we trying to interpolate down to days?

Reply:
Because the intervals are not always a clean 7 days. They are usually 7 days, but there's an occasional 6 or 8 thrown in. Interpolating down to days allows us to overlook this without worrying that 6-day intervals will have slightly smaller incident uptake (or 8-days slightly larger) than the customary 7-day interval would.

Reply:
Understood. How do you interpolate down to days? Is there some weird statistical artifact that would come from assuming constant uptake over a week?

The solution in the Bayesian world is nicer I think. We posit that there is an underlying (but unobserved) daily incident uptake, and the thing we observed is the cumulative uptake as of certain time points (or the incident uptake over some periods of time, which can be of arbitrary duration and different from one another)

Reply:
Linear interpolation. Given that the entire uptake curve is nonlinear, linear interpolation is surely "wrong," as you suspect. But on the few example data sets I've used, the posterior predictive checks look great, so I don't think it is "wrong" enough to matter.

Your suggested approach assumes a daily incident uptake: is it a problem that the uptake isn't linear over the course of the day? What if uptake peaks just before lunch and then again mid-afternoon? I see it like calculus - we can't ever quite escape the problem of nonlinearity in time, but we can squeeze it into smaller and smaller intervals until the problem is negligible. If 6-8 days is a small enough interval to make the problem negligible, I'm happy with the current approach. If we need smaller 1-day intervals, I'm happy to switch to your approach - I would just need to learn a bit more about it!

Reply:
I'm more concerned about statistical anomalies. Like, by going from 1 data point to 7, are you changing the confidence intervals somehow? Are you going to produce some weird edge effect (that won't be undone by summing daily uptakes into weeks after the analysis)?

This kind of interpolation will also be tricky when we do the v2 treatment, because although I know the observation error on the weekly data (ie the reported CIs), but I don't the error on the interpolated daily uptake

Reply:
1 data point is not becoming 7 data points; 1 data point stays 1 data point. That data point is simply normalized. If the incident uptake for the 7-day period ending on 2024-10-17 is 7%, then the daily average incident uptake for the 7-day period ending on 2024-10-17 is 1%. It's still a single data point on a single date - it's just normalized in a sense.